primitive/zk_rollup/src/merkle_circuit.rs

```diff
@@ -0,0 +1,173 @@
+use red_jubjub::sapling_hash;
+
+use self::root::check_membership;
+
+#[derive(Debug, PartialEq, Default)]
+pub struct BatchCircuit {
+    batch: Batch<Fr, Poseidon<Fr, 2>, 2, 2>,
+}
+
+impl BatchCircuit {
+    #[allow(dead_code)]
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(batch: Batch<Fr, Poseidon<Fr, 2>, 2, 2>) -> Self {
+        Self { batch }
+    }
+}
+
+impl Circuit<TatePairing> for BatchCircuit {
+    fn circuit(&self, composer: &mut Builder<TatePairing>) -> Result<(), Error> {
+        for RollupTransactionInfo {
+            transaction,
+            pre_root,
+            post_root,
+            pre_sender,
+            pre_receiver,
+            pre_sender_proof,
+            pre_receiver_proof,
+            post_sender_proof,
+            post_receiver_proof,
+        } in self.batch.transactions.iter()
+        {
+            let Transaction(sig, t) = transaction;
+
+            check_membership(
+                composer,
+                pre_sender.to_field_element(),
+                *pre_root,
+                pre_sender_proof.path,
+                pre_sender_proof.path_pos,
+            )?;
+
+            check_membership(
+                composer,
+                pre_receiver.to_field_element(),
+                *pre_root,
+                pre_receiver_proof.path,
+                pre_receiver_proof.path_pos,
+            )?;
+
+            check_signature(
+                composer,
+                t.sender_address.inner().into(),
+                *sig,
+                sapling_hash(&sig.r(), &t.sender_address.to_bytes(), &t.to_bytes()),
+            )?;
+
+            let post_sender = UserData {
+                balance: pre_sender.balance - t.amount,
+                ..*pre_sender
+            };
+
+            let post_receiver = UserData {
+                balance: pre_receiver.balance + t.amount,
+                ..*pre_receiver
+            };
+
+            check_membership(
+                composer,
+                post_sender.to_field_element(),
+                *post_root,
+                post_sender_proof.path,
+                post_sender_proof.path_pos,
+            )?;
+
+            check_membership(
+                composer,
+                post_receiver.to_field_element(),
+                *post_root,
+                post_receiver_proof.path,
+                post_receiver_proof.path_pos,
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use bls_12_381::Fr;
+    use ec_pairing::TatePairing;
+    use jub_jub::Fp;
+    use poly_commit::KzgParams;
+    use rand::rngs::StdRng;
+    use rand_core::SeedableRng;
+    use red_jubjub::{PublicKey, SecretKey};
+    use zero_plonk::prelude::*;
+    use zkstd::common::{CurveGroup, Group};
+
+    use crate::{
+        domain::{TransactionData, UserData},
+        operator::RollupOperator,
+        poseidon::Poseidon,
+    };
+
+    use super::BatchCircuit;
+
+    #[test]
+    fn batch_update() {
```

Comment:
rename to `zkrollup_test`

Reply:
This is not a zkrollup test, but it's better to rename the file I guess and the test to `batch_circuit_test`
The zkrollup one is is in the `lib.rs`

Reply:
Exactly